libs/server/Module/ModuleUtils.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Garnet.common;
+
+namespace Garnet.server
+{
+    public class ModuleUtils
+    {
+        public static bool LoadAssemblies(
+            IEnumerable<string> binaryPaths,
+            string[] allowedExtensionPaths,
+            bool allowUnsignedAssemblies,
+            out IEnumerable<Assembly> loadedAssemblies,
+            out ReadOnlySpan<byte> errorMessage)
+        {
+            loadedAssemblies = null;
+            errorMessage = default;
+
+            // Get all binary file paths from inputs binary paths
+            if (!FileUtils.TryGetFiles(binaryPaths, out var files, out _, [".dll", ".exe"], SearchOption.AllDirectories))
+            {
+                errorMessage = CmdStrings.RESP_ERR_GENERIC_GETTING_BINARY_FILES;
+                return false;
+            }
+
+            // Check that all binary files are contained in allowed binary paths
+            var binaryFiles = files.ToArray();
+            if (allowedExtensionPaths != null)
+            {
+                if (binaryFiles.Any(f =>
+                        allowedExtensionPaths.All(p => !FileUtils.IsFileInDirectory(f, p))))
+                {
+                    errorMessage = CmdStrings.RESP_ERR_GENERIC_BINARY_FILES_NOT_IN_ALLOWED_PATHS;
+                    return false;
+                }
+            }
+
+            // Get all assemblies from binary files
+            if (!FileUtils.TryLoadAssemblies(binaryFiles, out loadedAssemblies, out _))
+            {
+                errorMessage = CmdStrings.RESP_ERR_GENERIC_LOADING_ASSEMBLIES;
+                return false;
+            }
+
+            // If necessary, check that all assemblies are digitally signed
+            if (!allowUnsignedAssemblies)
+            {
+                foreach (var loadedAssembly in loadedAssemblies)
+                {
+                    var publicKey = loadedAssembly.GetName().GetPublicKey();
+                    if (publicKey == null || publicKey.Length == 0)
+                    {
+                        errorMessage = CmdStrings.RESP_ERR_GENERIC_ASSEMBLY_NOT_SIGNED;
```

Comment:
I wonder if there's any concerns about not unloading the assembly upon failure of signature verification? 🤔

Reply:
Good point.

The locations will have to be allowed by the admin for the assembly to be loaded through the 'LOAD MODULECS' command or the entire assembly specified using the --loadmodulecs argument. So, there is definitely admin control on what gets loaded. 

If the module initializer invocation cannot be precisely controlled, wondering would even explicitly unloading give any better security guarantees.

Reply:
Unloading wouldn't help in that case, you're correct. We could avoid loading the assembly for the check by using `System.Reflection.Metadata` to inspect the PE file manually:

 ```csharp
using var fs = File.OpenRead(modulePath);
using var peReader = new PEReader(fs);

var metadataReader = peReader.GetMetadataReader();
var assemblyPublicKeyHandle = metadataReader.GetAssemblyDefinition().PublicKey;

// if (assemblyPublicKeyHandle.IsNil) ..

var publicKeyBytes = metadataReader.GetBlobBytes(assemblyPublicKeyHandle);
```

But this also doesn't give any additional security guarantees AFAIK, as [.NET does not validate strong names at runtime](https://learn.microsoft.com/en-us/dotnet/standard/assembly/strong-named) anymore.

Reply:
Agreed. 
Updated to avoid loading assembly if signing check fails as a defensive measure.